patches/api/0336-Allow-components-in-conversation-prompts.patch

```diff
@@ -0,0 +1,107 @@
++            context.getForWhom().sendMessage(prefixComp.append(message));
++            // Paper end
+             if (!currentPrompt.blocksForInput(context)) {
+                 currentPrompt = currentPrompt.acceptInput(context, null);
+                 outputNextPrompt();
+diff --git a/src/main/java/org/bukkit/conversations/Prompt.java b/src/main/java/org/bukkit/conversations/Prompt.java
+index fcca208c0f31b41ab67323d0e77a8e3ecf9e78dd..f0f2762e52e8f65161c6edf1b378a9a39d636a95 100644
+--- a/src/main/java/org/bukkit/conversations/Prompt.java
++++ b/src/main/java/org/bukkit/conversations/Prompt.java
+@@ -25,7 +25,16 @@ public interface Prompt extends Cloneable {
+      * @return The text to display.
+      */
+     @NotNull
+-    String getPromptText(@NotNull ConversationContext context);
++    // Paper start - Allow components in conversation prompts
++    default String getPromptText(@NotNull ConversationContext context) {
++        throw new UnsupportedOperationException("Please override either getPromptText or getPromptMessage");
++    }
++
++    @NotNull
++    default net.kyori.adventure.text.Component getPromptMessage(@NotNull ConversationContext context) {
++        return io.papermc.paper.text.PaperComponents.legacySectionSerializer().deserialize(getPromptText(context));
++    }
++    // Paper end
+ 
+     /**
+      * Checks to see if this prompt implementation should wait for user input
+diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
+index 3e8cd3971ac8256a40d9b85cd7514998c965512c..d44859f7d4bcf4cac31528826de3d967fa07d9cb 100644
+--- a/src/main/java/org/bukkit/entity/Player.java
++++ b/src/main/java/org/bukkit/entity/Player.java
+@@ -2295,4 +2295,11 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
+     @Override
+     Spigot spigot();
+     // Spigot end
++
++    // Paper start - Allow components in conversation prompts
++    @Override
++    default void sendMessage(final @NotNull Component message) {
++        this.sendMessage(net.kyori.adventure.identity.Identity.nil(), message);
++    }
++    // Paper end
+ }
+diff --git a/src/test/java/org/bukkit/conversations/FakeConversable.java b/src/test/java/org/bukkit/conversations/FakeConversable.java
+index 567bfd7d3e45d83b5520af6ddd0140c2b84139b2..c5036858671b6d346bcb9db465f7ef3ea13dfe46 100644
+--- a/src/test/java/org/bukkit/conversations/FakeConversable.java
++++ b/src/test/java/org/bukkit/conversations/FakeConversable.java
+@@ -51,6 +51,13 @@ public class FakeConversable implements Conversable {
+         lastSentMessage = message;
+     }
+ 
++    // Paper end - Allow components in conversation prompts
++    @Override
++    public void sendMessage(@NotNull net.kyori.adventure.text.Component component) {
++        lastSentMessage = ((net.kyori.adventure.text.TextComponent)component).content();
```

Comment:
I guess this is good enough for a test but really, a serialisier should be used in situations like this (plain or legacy I guess)

Reply:
Fair enough, I just wanted the most simple thing here given that that no test actually gets to this spot. Would have to resort to referencing that directly though, since calling `plainSerializer()` would fail here. When I do that, I'll expand the tests while I'm at it.